File: 907-koko-eating-bananas/koko-eating-bananas.java
class Solution {
    public int minEatingSpeed(int[] piles, int h) {
        int right = 0;
        for(int pile : piles)
            right = Math.max(right, pile);
        int left = 1;
        while(left<= right) {
            int mid = left + (right - left) / 2;
            long totalHours = totalHours(piles, mid);
            if(totalHours <= h) {
                right = mid - 1;
            } else {
                left = mid +1;
            }
        }
        return left;
    }

    long totalHours(int[] piles, int h) {
        long totalHours = 0;
        for(int i=0;i<piles.length;i++) {
            totalHours += (piles[i] + h - 1) / h;
        }
        return totalHours;
    }
}

File: 907-koko-eating-bananas/koko-eating-bananas.py
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        right = max(piles)
        
        left = 1
        res = right

        while left <= right:
            mid = (left + right) // 2
            totalHours = 0
            for i in piles:
                totalHours += math.ceil(i/mid)

            if totalHours <= h:
                res = min(res, mid)
                right = mid - 1
            else:
                left = mid + 1

        return res

